h2/src/main/org/h2/table/TableFilter.java

```diff
@@ -1158,14 +1159,8 @@ public void lockRowAdd(ArrayList<Row> rows) {
      *
      * @param forUpdateRows the rows to lock
      */
-    public void lockRows(ArrayList<Row> forUpdateRows) {
-        for (Row row : forUpdateRows) {
-            Row newRow = row.getCopy();
-            table.removeRow(session, row);
-            session.log(table, UndoLogRecord.DELETE, row);
-            table.addRow(session, newRow);
-            session.log(table, UndoLogRecord.INSERT, newRow);
-        }
+    public void lockRows(Iterator<Row> forUpdateRows) {
```

Comment:
While I absolutely agree with your reasoning about optimization opportunities expansion, I tend to hold to KISS and "avoid premature optimization" principles. Signatures with specific implementations, where interface would suffice, definitely violate first one, and the fact that this call is made once per SQL statement... On the other hand, you are right that passing around Iterator is not common, so how about Iterable, and let JIT to figure out that there is only a single call site with ArrayList as an argument?

Reply:
C2 (not sure about C1) is usually smart enough to determine that passed interface is always or in many cases have exactly the same real data type and provide optimized code for it (with slow fallback code for generic case). Interpreter is not so advanced, index-based loop is a fastest way for `ArrayList` in interpreted code and interfaces are usually slower than classes.

Sometimes C2 can inline everything and in this case there is no difference what exactly will be passed, `ArrayList`, `Iterable`, or `Iterator`. But when this is not possible (different implementations of called method exist) deviations in performance appears.

From my point of view it's OK to use `Iterable` here, but `Iterator` is not reasonable. It's not very usable, because enhanced `for` loops cannot be used with it.